R/R/prophet.R

```diff
@@ -1274,9 +1288,134 @@ fit.prophet <- function(m, df, ...) {
 #' forecast <- predict(m, future)
 #' plot(m, forecast)
 #' }
-#'
 #' @export
 predict.prophet <- function(object, df = NULL, ...) {
+  
+  if (is.null(df)) {
+    m <- object
+    history.and.future <- m$history
+  } else {
+    if (nrow(df) == 0) {
+      stop("Dataframe has no rows.")
+    }
+    out <- setup_dataframe(object, df)
+    history.and.future <- out$df
+    m <- out$m
+  }
+  
+  out2 <- make_all_seasonality_features(m, history.and.future)
+  seasonal.features <- out2$seasonal.features
+  component.cols <- object$train.component.cols
+  
+  # Construct input to stan
+  dat <- list(
+    T = nrow(object$history),
+    K = ncol(seasonal.features),
+    S = length(object$changepoints.t),
+    y = object$history$y_scaled,                         
+    t = object$history$t,                                
+    cap = rep(0, nrow(object$history)),                         
+    t_change = array(object$changepoints.t),
+    X = as.matrix(seasonal.features[1:nrow(object$history),]),            
+    sigmas = array(out2$prior.scales),
+    tau = object$changepoint.prior.scale,
+    trend_indicator = as.numeric(object$growth == 'logistic'),
+    s_a = array(component.cols$additive_terms),
+    s_m = array(component.cols$multiplicative_terms),
+    
+    T_pred = nrow(history.and.future),
+    t_pred = history.and.future$t,
+    cap_pred = 
+      if(class(history.and.future$cap) == "NULL") { 
+        rep(0, length(history.and.future$t)) 
+      } else { 
+        history.and.future$cap 
+      },
+    X_pred = as.matrix(seasonal.features),
+    n_samp = object$uncertainty.samples
+  )
+  
+  # Future # of change points assuming a homog. Poisson process
+  dat$S_pred <- stats::qpois(0.95, dat$S * (dat$T_pred - dat$T) / dat$T )
```

Comment:
the notation on the Stan side is quite poor, which is my fault because I'm the one who originally defined it :-) But, T_pred and T are the number of rows. What we really want here are the relative lengths of the periods in "real" time, which may be different if the times in the history or future are not regularly spaced. That is, we want little `t` (which is the time) and not the big `T` (which is number of rows). `t` for the history is mapped to [0, 1]. The future begins at 1, and then goes forward from there for however long the future is relative to the history. (e.g., if we have 2 years of data and make a prediction of 1 year, dat$t will be on [0, 1] and dat$t_pred will be on [1, 1.5]).

So what we want here is `dat$S * (max(dat$t_pred) - 1) / (1 - 0) = dat$S * (max(dat$t_pred) - 1)`. Does that make sense?

Reply:
Got it, I mixed up upper and lower case t.